### ISSUE TITLE
[rcore] KeyboardKey only supports American keyboard 

### ISSUE THREAD

**Issue by PizzaGame01** [raysan5/raylib#4019]
I'm using a Europe (German) keyboard.
Its happend in [Raylib-cs](https://www.nuget.org/packages/Raylib-cs), but becaus its a port of Raylib i think this problem is still in original Raylib.

If i do this:
```cs
if (Raylib.IsKeyDown(KeyboardKey.Z))
{
    Console.WriteLine("Called!");
}
```
Its happen nothing by pressing Z, but if i press Y the if statement is executing.

This problem is also by ``int key = Raylib.GetKeyPressed();``

```cs
int key = Raylib.GetKeyPressed();
Console.WriteLine((KeyboardKey)key).ToString());
```

Clicked: Y
Output: Z

**Comment by MrScautHD** [comment]
I can agree.

**Comment by JupiterRider** [comment]
I think this is behavior intended. See https://github.com/raysan5/raylib/issues/2822

**Comment by raysan5** [comment]
@PizzaGame01 I'm afraid there is not an easy solution to this problem, the keyboard mapping should be managed at GLFW, that raylib uses or at user side. Do you know of another possible solution?

**Comment by raysan5** [comment]
If anyone finds this issue and knows of a simple multi-platform solution to address it, please, comment

**Comment by MrScautHD** [comment]
> @PizzaGame01 I'm afraid there is not an easy solution to this problem, the keyboard mapping should be managed at GLFW, that raylib uses or at user side. Do you know of another possible solution?

has SDL 2 the same problem?

**Comment by JeffM2501** [comment]
The problem statement in this issue is flawed, that is why there is confusion.

The keyboard key (KEY_) enum is not the keyname, it's the physical keycode. That enum was named after the American layout, because the enum has to be named something.

The enum can not and should not change based on keyboard layout. it represents physical buttons.
A FPS game on a dvorak keyboard would not use WASD, it'd use ,AEO, because they are in the same physical location.
A raylib game should not be expected to know the layout when checking for physical keys, it should simply check for KEY_W, KEY_A, KEY_D, KEY_S. 
Text input is mapped using GetCharPressed and is handled correctly.
What we are missing is a way to get the key name from the platform so that a game can show the mapped key name in things like settings and control bindings.

There should be an API where the user can ask "What is the name of KEY_W", and on dvorak it would return ',', or whatever the characater code is for the keyboard mapping.

GLFW exposes this via glfwGetKeyName, and I'm sure that other platforms do as well.
https://www.glfw.org/docs/3.3/group__input.html#gaeaed62e69c3bd62b7ff8f7b19913ce4f

We should add a` const char* GetKeyName(KeyboardKey key) `function to raylib to handle this and let each platform implement it.

This will solve the core issue and provide access to all 3 levels of keyboard info for keys

- Physical Key (KeyboardKey  enum)
- Typed character (GetCharPressed results)
- KeyName (new GetKeyName function)

**Comment by MrScautHD** [comment]
@raysan5 can you may reopen it?

**Comment by Mossieur-Patate** [comment]
The function GetKeyName is not found because its lacks a prototype in raylib.h. I tried adding a prototype locally and it worked, so I am trying to fix it. (Just created an account on GitHub just to help, I wish Raylib were on GitLab.)
`RLAPI const char *GetKeyName(int key);`
By the way, I also notice I should be adding it to [raylib_to_parse.h](https://github.com/raysan5/raylib/blob/79188f570bf15791c3abad619933d039448023b1/projects/Notepad%2B%2B/raylib_npp_parser/raylib_to_parse.h) to get Notepad++ to suggest the function, apparently; tell me if I forgot any other file concerned.

The function is missing from the main cheatsheet, so it should also be added there.

By the way, it does not seem simple to simulate the layout-independent equivalents for the other functions of this kind, so I guess there should also be an `IsCharPressed` function and so on (I am especially thinking of `IsCharDown` for player movement). Current signatures are as follows.
```
RLAPI bool IsKeyPressed(int key);                             // Check if a key has been pressed once
RLAPI bool IsKeyPressedRepeat(int key);                       // Check if a key has been pressed again
RLAPI bool IsKeyDown(int key);                                // Check if a key is being pressed
RLAPI bool IsKeyReleased(int key);                            // Check if a key has been released once
RLAPI bool IsKeyUp(int key);                                  // Check if a key is NOT being pressed
RLAPI int GetKeyPressed(void);                                // Get key pressed (keycode), call it multiple times for keys queued, returns 0 when the queue is empty
RLAPI int GetCharPressed(void);                               // Get char pressed (unicode), call it multiple times for chars queued, returns 0 when the queue is empty
RLAPI void SetExitKey(int key);                               // Set a custom key to exit program (default is ESC)
```

Two related-yet-more-general suggestions I put here (feel free to put this elsewhere):

1. Maybe a systematic test through some script could check if all implemented functions do have their matching prototypes in the right header file. This would have prevented `GetKeyName` to miss its own.
2. Maybe add super quick guidelines in [CONTRIBUTING.md](https://github.com/raysan5/raylib/blob/79188f570bf15791c3abad619933d039448023b1/CONTRIBUTING.md) to pinpoint where prototypes should be added. Same argument, plus I found raylib_to_parse.h by chance looking for any instance of ‘raylib.h’, although I do not use Notepad++ but Sublime Text, and thus would have probably not had a clue about the error otherwise. I am unaware if there is already some test that checks whether raylib_to_parse.h matches raylib.h, but I guess this could be useful.

(Sorry if these latter suggestions are redundant, just my two cents from my quick experience of this repository.)
